File: LEETCODE/MEDIUM/DAY 25/PYTHON SOLUTION/Brute Force (O(n²)).py
class Solution:
    def isValidBST(self, root):
        if not root:
            return True
        return (self.isSubtreeLess(root.left, root.val) and
                self.isSubtreeGreater(root.right, root.val) and
                self.isValidBST(root.left) and
                self.isValidBST(root.right))

    def isSubtreeLess(self, node, val):
        if not node: return True
        return node.val < val and self.isSubtreeLess(node.left, val) and self.isSubtreeLess(node.right, val)

    def isSubtreeGreater(self, node, val):
        if not node: return True
        return node.val > val and self.isSubtreeGreater(node.left, val) and self.isSubtreeGreater(node.right, val)

File: LEETCODE/MEDIUM/DAY 25/JAVA SOLUTION/Brute Force (O(n²)).java
public boolean isValidBST(TreeNode root) {
    if (root == null) return true;
    return isSubtreeLessThan(root.left, root.val)
        && isSubtreeGreaterThan(root.right, root.val)
        && isValidBST(root.left)
        && isValidBST(root.right);
}

private boolean isSubtreeLessThan(TreeNode node, int val) {
    if (node == null) return true;
    return node.val < val 
        && isSubtreeLessThan(node.left, val)
        && isSubtreeLessThan(node.right, val);
}

private boolean isSubtreeGreaterThan(TreeNode node, int val) {
    if (node == null) return true;
    return node.val > val
        && isSubtreeGreaterThan(node.left, val)
        && isSubtreeGreaterThan(node.right, val);
}
